src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
+                let future = self.tx.clone().send(ServerMessage::Shutdown);
+                let me = self.clone();
+                return Box::new(future.then(move |_| {
+                    let mut shutting_down = ShuttingDown::new();
+                    shutting_down.set_stats(me.get_stats());
+                    res.set_shutting_down(shutting_down);
+                    Ok(Message::WithoutBody(res))
+                }))
+            } else {
+                warn!("handle_client: unknown command");
+                res.set_unknown(UnknownCommand::new());
+            }
+
+            future::ok(Message::WithoutBody(res)).boxed()
+        }
+    }
+}
+
+impl<C> SccacheService<C>
+    where C: CommandCreatorSync,
+{
+    pub fn new(storage: Arc<Storage>,
+               handle: Handle,
+               pool: CpuPool,
+               tx: mpsc::Sender<ServerMessage>,
+               info: ActiveInfo) -> SccacheService<C> {
+        SccacheService {
+            stats: Rc::new(RefCell::new(ServerStats::default())),
+            storage: storage,
+            compilers: Rc::new(RefCell::new(HashMap::new())),
+            force_recache: env::var("SCCACHE_RECACHE").is_ok(),
+            pool: pool,
+            handle: handle,
+            creator: C::new(),
+            tx: tx,
+            info: info,
+        }
+    }
+
+    /// Get stats about the cache.
+    fn get_stats(&self) -> CacheStats {
+        let mut stats = CacheStats::new();
+        let mut stats_vec = self.stats.borrow().to_cache_statistics();
+
+        let mut stat = CacheStatistic::new();
+        stat.set_name(String::from("Cache location"));
+        stat.set_str(self.storage.location());
+        stats_vec.insert(0, stat);
+
+        for &(s, v) in [("Cache size", self.storage.current_size()),
+                       ("Max cache size", self.storage.max_size())].iter() {
+            v.map(|val| {
+                let mut stat = CacheStatistic::new();
+                stat.set_name(String::from(s));
+                stat.set_size(val as u64);
+                stats_vec.insert(0, stat);
+            });
+        }
+
+        stats.set_stats(RepeatedField::from_vec(stats_vec));
+        stats
+    }
+
+    /// Zero and return stats about the cache.
+    fn zero_stats(&self) -> CacheStats {
+        *self.stats.borrow_mut() = ServerStats::default();
+        self.get_stats()
+    }
+
+
+    /// Handle a compile request from a client.
+    ///
+    /// This will either start compilation and set a `CompileStarted`
+    /// response in `res`, or set an `UnhandledCompile` response in `res`.
+    fn handle_compile(&self, mut compile: Compile)
+                      -> Box<Future<Item = SccacheResponse, Error = io::Error>>
+    {
+        let exe = compile.take_exe();
+        let cmd = compile.take_command().into_vec();
+        let cwd = compile.take_cwd();
+        // See if this compiler is already in the cache.
+        match self.compiler_info_cached(&exe) {
```

Comment:
Feels like this could be rewritten so the cache lookup just always returned a `Future` to simplify the logic, like:
`self.get_compiler_info(&exe).then(|info| self.check_compiler(info, cmd, cwd))`

When I wrote that code originally I was actually wishing for some sort of `Future`-like abstraction!